### PR TITLE
fix: fix copycat orphan detection on cron and OOM

### PR DESCRIPTION
**Pull request by shyba**
Current cron gets blocks through a cache by height, so an orphan block replacing an indexed block isn't seen. This changes it so the default is to bypass the block cache, thus overwriting orphan blocks.

In the process of testing, an OOM was triggered and fixed. Caused by reading the wrong offset and interpreting the bad chunk as an integer for the bundle size. Which later causes a huge allocation of a list. 

The other fixes have tests. They guard against orphan txs during reindex.

### CODE REVIEW

File: src/dev_copycat_arweave.erl [hunk 1]
```diff
@@ -114,17 +119,25 @@ list_index_blocks(Current, To, Opts, Acc) ->
...
+                hb_ao:resolve(
+                    <<
+                        ?ARWEAVE_DEVICE/binary,
+                        "/block=",
+                        (hb_util:bin(Height))/binary
+                    >>,
+                    Opts
+                );
+            false ->
+                dev_arweave:fetch_block({height, Height}, Opts)
```
Comment (JamesPiechota) [inline src/dev_copycat_arweave.erl]: I'm a little worried about all the different permutations we're introducing for querying a block. e.g. I believe there are now 4 ways to get a block depending on the values of `arweave_index_blocks` and `block_cache`

TBH I don't think the block cache is used at all anywhere (at least not intentionally), so I'm not sure how important it is that we preserve it. Especially given all the trouble it is causing. But Sam will need to weigh in on that.

If we want to keep the block cache I'm thinking something like this might work best:
- remove `block_cache` and `arweave_index_blocks` and replace with a single `arweave_cache_blocks = true|false`
- everywhere that queries the block outside of `dev_arweave` uses the `hb_ao:resolve` API and doesn't call `dev_arweave:xxx` directly (I believe this is the desired way of handling inter-device coupling)
- whenever a block is queried, `dev_arweave` consults `arweave_cache_blocks` if `true` *and if* the block is more than 19 blocks old: it is cached

This should address all the issues while preserving the cross-device coupling model, and controlling the cache with a single variable.

That said it does depend on `dev_arweave` always knowing the current tip height which I'm not sure it does.

Another alternative is just to set `arweave_index_blocks` to default to false. Ultimately that's how we intended all of the current production nodes to run and making that small change should address the rolling index fork recovery issue you noticed.

File: src/hb_store_arweave.erl [hunk 2]
```diff
@@ -201,19 +209,29 @@ load_tx(ID, StartOffset, Length, Opts) ->
...
-                {error, Reason} ->
-                    {error, Reason}
+            case Length of
+                0 ->
+                    {ok, hb_message:convert(
+                        TXHeader, <<"structured@1.0">>, <<"tx@1.0">>, Opts)};
+                _ ->
+                    case read_chunks(StartOffset, Length, Opts) of
+                        {ok, Data} ->
+                            TXWithData = ar_tx:generate_chunk_tree(
```
Comment (JamesPiechota) [inline src/hb_store_arweave.erl]: I believe this check should be in dev_codec_tx. If a transaction's data does not match its `data_root` then is should be considered invalid.

If the check isn't in `dev_codec_tx` I think it's just cause we deferred adding it due to time constrained. I know we had deferred adding it during the first pass in June last year, but can't remember if we ever returned to add it.

If it's missing, it should probably be added somewhere in `dev_codec_tx` so that we have a consistent definition of a valid L1 transaction. Perhaps in the verify logic or maybe the normalize logic

File: src/dev_arweave.erl [hunk 3]
```diff
@@ -513,6 +513,12 @@ fetch_and_collect(Offsets, Opts) ->
 
 %% @doc Generate a list of offsets from Start to End (inclusive) stepping by
 %% Step bytes. Used to produce candidate query offsets at 256KiB increments.
+%% Capped at 1M items (~256GB at 256KiB step). Ranges above this would
+%% require streaming chunk assembly, which is not currently supported.
+-define(MAX_OFFSET_LIST, 1_000_000).
+generate_offsets(Start, End, Step)
```
Comment (JamesPiechota) [inline src/dev_arweave.erl]: I can't remember why we had Step be configurable. If you're adding a maximum size, probably best to remove `Step` and hard code it `DATA_CHUNK_SIZE`. Then the maximum value is expliciltly a size cap (Right now with Step being configurable a 1_000_000 cap might not always make sense - e.g. if Step is 1. But ultimately I don't think Step needs to be confgurable? I can't remember)

And so I have full context: did you hit the OOM in practice? Or was it only in a test? i.e. could this explain some of the mysterious memory issues we've been having? or is this primarily defensive programming?

File: src/dev_copycat_arweave.erl (hunk 1 above)
Comment (shyba) [reply src/dev_copycat_arweave.erl]: reverted that, changed the default `arweave_index_blocks` instead

File: src/dev_arweave.erl (hunk 3 above)
Comment (shyba) [reply src/dev_arweave.erl]: removed, good point

File: src/dev_arweave_offset.erl [hunk 4]
```diff
@@ -207,30 +207,34 @@ read_remaining_data(StartOffset, HeaderSize, PrefixSize, Length, Opts) ->
...
 
-%% @doc Recover the global start offset of the containing bundle from the end
-%% offset of the chunk in global space and its end offset inside the bundle.
-bundle_start_offset(ChunkJSON) ->
+%% @doc Recover the global start offset and total size of the containing
+%% bundle from the chunk metadata.
+bundle_start_and_size(ChunkJSON) ->
     AbsEnd = hb_util:int(maps:get(<<"absolute_end_offset">>, ChunkJSON)),
-    ChunkEndInBundle =
+    BundleSize =
```
Comment (JamesPiechota) [inline src/dev_arweave_offset.erl]: `ar_merkle:extract_note` always confuses me - are you sure it will yield the bundle size? It's not obvious to me - maybe add a comment explaining why the note represents the size in this case?

File: src/hb_store_arweave.erl (hunk 2 above)
Comment (shyba) [reply src/hb_store_arweave.erl]: I think when we call `load_tx` and give it both an `ID` and an `Offset` we are telling it to load a tx with that ID on that offset. If the loaded tx from there is not the one we asked, its either caller or `load_tx` the place to handle it.

`dev_codec` would be nice, but it would add a very high cost and would not catch this case. If I ask for the wrong ID and `load_tx` goes to the wrong place and loads a perfectly valid transaction, the only place with full context is either this or the caller.

Comment (JamesPiechota) [reply src/hb_store_arweave.erl]: This check is ultimately something we deferred from the original `dev_codec_tx` implementation because it introduced some complexity we didn't have capacity to deal with then i.e. handling the case where a new unsigned TX is being created and you want to initialize `data_root` from the data vs, the case where you receive a signed TX and want to validate `data_root` against the `data` *iff* data is present - definitely solvable, but introduced enough permutations that we punted on it.

 I think the places to implement it are wherever we call `dev_arweave_common:normalize` - this happens in both `to` and `from`
- [from](https://github.com/permaweb/HyperBEAM/blob/neo/edge/src/dev_codec_tx.erl#L79)
- [to](https://github.com/permaweb/HyperBEAM/blob/neo/edge/src/dev_codec_tx.erl#L146)

Right now `normalize` builds the chunk tree and sets `data_root`. The right logic is probably to compare against `data_root` before overwriting it, if data_root is set and different, then probably throw an error or otherwise flag the TX as invalid. There may be other bits of code that need to be updated as well.

Alternatively we could leave this logic here where you have it, but in this case I believe the argument would be:
- this issue needs an urgent fix and we don't have time to fix it "right"
- due to the severity and urgency of the issue we think incurring tech debt by implementing the patch here rather than in `dev_codec_tx` is warranted

I'm unclear on the urgency/severity of the issue. My preference would be to add this logic to the codec (where all our other data validation happens) - but then again I'm the one who punted on adding it in the first place, so I realize the risk in delaying a fix is that it may be delayed further.

Comment (shyba) [reply src/hb_store_arweave.erl]: moved, but I could not do the same to the ans104 check. Does the same apply there?

File: src/dev_arweave_offset.erl (hunk 4 above)
Comment (shyba) [reply src/dev_arweave_offset.erl]: Yes, that confused me as well. My logic is that usually `start = end - size` and we had `AbsEnd - ChunkEndInBundle.` being returned as the start offset, so the rename matched. I found more context around and updated the comment. Let me know if it looks correct now

Comment (JamesPiechota) [reply src/dev_arweave_offset.erl]: It might be worth adding a test for this, as I'm still not sure it's correct. The ChunkJSON provided is just for the first chunk in the bundle - that's why the old logic worked - we take the absolute end offset of the chunk and subtract the size of the chunk and we're left with the start of the bundle.

I'm not sure this allows us to infer the size of the bundle. I could see it being encoded in the data_path somehow (e.g. the right bound of the root node in the merkle tree should be the rightbound of the TX (maybe?), but I'm not sure whether that data makes it into all data_paths - my guess is it doesn't and that the note in this case is the right bound of the first chunk)

But if you've tested and confirmed this in fact the size of the bundle, it's definitely possible i've gotten it wrong.

Comment (shyba) [reply src/dev_arweave_offset.erl]: It was confusing to me as well and the path isnt critical, so I removed there. It covers less ground but the main one is still covered.

File: src/dev_arweave.erl [hunk 5]
```diff
@@ -944,6 +954,26 @@ event_request(Path, Method, Status, Extra) ->
...
+    Garbage = <<255:(32*8)>>,
+    HeaderSize = ar_bundles:bundle_header_size(Garbage),
+    TXDataSize = 10_000_000,
+    ?assert(HeaderSize > TXDataSize),
+    ?assertEqual(
+        {error, invalid_bundle_header},
+        bundle_header_with_size_check(HeaderSize, TXDataSize)
+    ).
+
+bundle_header_with_size_check(HeaderSize, MaxSize)
```
Comment (JamesPiechota) [inline src/dev_arweave.erl]: is this only used in the test above? If so I think this check covers it: https://github.com/permaweb/HyperBEAM/pull/781/changes#diff-a6acc62ee3f34a6566c482abad60bca6956207fc45b65ba1e80ed753aa7a84e5R964

File: src/dev_arweave_common.erl [hunk 6]
```diff
@@ -216,12 +216,25 @@ reset_owner_address(TX) ->
 
 normalize_data_root(Item = #tx{data = Bin, format = 1})
         when is_binary(Bin) andalso Bin =/= ?DEFAULT_DATA ->
-    Item#tx{data_root = ar_tx:data_root(legacy, Bin)};
+    NewRoot = ar_tx:data_root(legacy, Bin),
```
Comment (JamesPiechota) [inline src/dev_arweave_common.erl:219]: This all looks about right for me, but without the signed/unsigned check and some other possible edge cases, I suspect it might fail sometimes. Do all the tests pass? if so that's a good sign.

The failure mode I might watch out for are some of the roundrip tests (Msg -> Tx -> Msg and Tx -> Msg -> Tx) particularly for the unsigned test cases. As long as a TX is unsigned it's probably acceptable to update and overwrite the data_root as you build up the TX.

Comment (shyba) [reply src/dev_arweave_common.erl:219]: they do but I have around 17 of them that never pass because of my slower connection, so it would be good to run it before merge

File: src/dev_arweave.erl (hunk 5 above)
Comment (shyba) [reply src/dev_arweave.erl]: great finding. This was testing another code I deleted and became nonsense after a refactor. Removed.

**Review by JamesPiechota — APPROVED** [review]

**Comment by shyba** [comment]
#817 replaced it
